utils/parametric/spec/trace.py

```diff
@@ -154,107 +152,49 @@ def decode_v06_stats(data: bytes) -> V06StatsPayload:
+    span_id = int(span_id)
+    for span in trace:
+        if span.get("span_id") == span_id:
+            return span
+    raise AssertionError(f"Span with id={span_id} not found. Trace={trace}")
 
-    max_similarity = -math.inf
-    max_similarity_span = trace[0]
-    for other_span in trace:
-        similarity = _span_similarity(span, other_span)
-        if similarity > max_similarity:
-            max_similarity = similarity
-            max_similarity_span = other_span
-    return max_similarity_span
 
+def find_span_in_traces(traces: List[Trace], trace_id: int, span_id: int) -> Span:
+    """Return a span from a list of traces by `trace_id` and `span_id`."""
+    trace = find_trace(traces, trace_id)
+    return find_span(trace, span_id)
 
-def find_span(trace: Trace, span: Span) -> Span:
-    """Return a span from the trace matches all fields in `span`."""
-    return _assert_span_match(span, find_similar_span(trace, span))
 
+def find_only_span(traces: List[Trace]) -> Span:
+    """Return the only span in a list of traces. Raises an error if there are no traces or more than one span."""
+    assert len(traces) == 1, traces
+    assert len(traces[0]) == 1, traces[0]
+    return traces[0][0]
 
-def find_similar_span_in_traces(traces: List[Trace], span: Span) -> Span:
-    """Return a span from the traces which most closely matches `span`."""
-    assert len(traces) > 0
 
-    max_similarity = -math.inf
-    max_similarity_span = None
-    for trace in traces:
-        similar_span = find_similar_span(trace, span)
-        if max_similarity_span is None:
-            max_similarity_span = similar_span
-        similarity = _span_similarity(span, max_similarity_span)
-        if similarity > max_similarity:
-            max_similarity_span = similar_span
-            max_similarity = similarity
-    return max_similarity_span
-
-
-def find_span_in_traces(traces: List[Trace], span: Span) -> Span:
-    """Return a span from the traces that matches all fields in `span`."""
-    return _assert_span_match(span, find_similar_span_in_traces(traces, span))
-
-
-def _assert_span_match(span: Span, similar: Span) -> Span:
-    for var in Span.__annotations__:
-        if not var.startswith("__"):
-            val = span.get(var)
-            s_val = similar.get(var)
-            if val is not None and val != s_val:
-                # TODO remove this special handling once nodejs sets service in the same way.
-                # Right now, nodejs sets a tag named "service" in meta with the correct value.
-                if var == "service":
-                    meta = similar.get("meta")
-                    if meta is not None:
-                        s_val = meta.get(var)
-                assert (
-                    val == s_val
-                ), f"Span field '{var}' mismatch '{val}' != '{s_val}'\nSpan   : {span}\nSimilar: {similar}"
-    return similar
+def find_chunk_root_span(trace: Trace) -> Span:
```

Comment:
This `find_chunk_root_span` function is currently mixing up a few concepts (root spans and trace-level tags) that do not have a consistent correlation that we can rely on.

### Root span

The root span of a trace is the span with no (local or remote) parent span. The correct way to find the root span of a trace is to find the span where `parent_id is null or 0`. A local trace must have exactly one root span. If a local trace is split into multiple chunks, only one of the chunks will contain the root span. In other words, a trace chunk can only ever have zero or one root spans.

### Trace-level tags

For several reasons, we are unfortunately very inconsistent with how we treat "trace-level" tags. We cannot rely on these tags to find the root span.
- `_dd.origin` tag is added to all spans in a trace.
- `language` tag is added to multiple spans in a trace. Originally, we added it to all spans except spans that represent an external service call. Today we can safely add it to all spans. It was never added only to the root span.
- `runtime-id` is added to all top-level spans in a trace. The root span will always have this tag, since it is always a top-level span, but other spans in the same trace can also have this tag.
- `_dd.p.tid` and other propagated tags (`_dd.p.*`) are added to the first span in every trace chunk. It is not guaranteed that this first span will be the root span, hence propagated tags have no relationship to the root span.

Reply:
There are other examples:
- `process_id` is added only to root spans, so technically is could be used to find the root span. but really, there's no reason to rely on span tags when we can use the `parent_id` which is the definition of roto span and always works. (for example, the behavior of `process_id` could change in the future) 
- "priority sampling rate" tags (`_dd.*_psr`) are added to the root span as well
- `_sampling_priority_v1` has changed over time and is pretty complicated. Let's just say that at a minimum, it needs to show up in at least one span (any span) in every trace chunk. But for the Trace Agent, placing it in some spans is more efficient than others.
- `env` is added to all spans
- `version` is added to all spans whose service name is the same as the default service name

Reply:
With this change, I will go over tests that use `find_trace_with_trace_level_tags` and make sure this helper is used correctly.

Reply:
`find_trace_with_trace_level_tags` still doesn't make sense to me. First, shouldn't it be `find_span...`? You're trying to find and return a single span, not a trace.

But more importantly, different "trace-level" tags have different rules about which spans we add them to. I listed 8 examples in my comment above. So you cannot find a single spans that contains all the trace-level tags. Tests for each individual tag should know where to look for the span and assert that they were added to the correct span.

Maybe we have different definitions of what a "trave-level" tags is, since that's more of an implicit concept that we haven't defined anywhere in APM.

Reply:
After talking over this on zoom I understand better now, thanks! `find_trace_with_trace_level_tags` was looking for the span with `_dd.p.*` tags (aka propagated tags).

We decided it would be better to grab the first span in the trace chunk, since that's where those tags are supposed to be. Searching for any span with the tag can result in test passing that should fail (if the tag is on the wrong span).